WIP: Allow tagging of s3 bucket and dynamo db

BernhardBln opened the pull request:
This is WIP - no tests were created for this yet, but the tagging itself works already.

Please give some feedback if that looks ok, then I will start adding some tests too.

And remember I'm totally new to go...

**Note**: I decided against putting the tags outside of config, but inside of remote_state, for two reasons:

1. (Minor) - I was not able to get access to the whole remote_state object in the remote_state_s3 file
2. (Major) - I think it does not make sense to have an option which is s3-specific top-level in remote_state - if you ever add another provider, you would have to explain that these two config items can be set only if the backend is s3, which I find worse than putting it inside of config and filtering out later (which I do in two places, when reading the config in remote_state_s3, and when iterating over the config items to prepare the command line options for terraform)

If you still think I should place them outside of remote_state, I will change that, but I find the second argument quite strong.

So configuration looks like this now:

```
terragrunt = {

  remote_state {
    backend = "s3"
    config {
      bucket = "..."
      [...]

      s3_bucket_tags {
        owner = "team name"
        name = "Terraform state storage"
      }

      dynamotable_tags {
        owner = "team name"
        name = "Terraform lock table"
      }

    }
  }
}
```

You can set arbitrary tags (just follow the limitations of AWS, I do not check e.g. if a tag value is too long etc.

In case the tag mappings are missing, nothing is tagged (when they are there but are empty, an empty tagging is made, which could be prevented if you think that's necessary). 

In case they appear multiple times (like, if s3_bucket_tags appears twice or more), an error is printed and we stop.

brikis98 commented on dynamodb/dynamo_lock_table.go:
```diff
@@ -99,7 +99,44 @@ func CreateLockTable(tableName string, readCapacityUnits int, writeCapacityUnits
+	return nil
+}
+
+func tagTableITagsGiven(tags map[string]string, createTableOutput *dynamodb.CreateTableOutput, client *dynamodb.DynamoDB, terragruntOptions *options.TerragruntOptions) error {
```
s/tagTableITagsGiven/tagTableIfTagsGiven

BernhardBln replied:
Done

brikis98 commented on dynamodb/dynamo_lock_table.go:
```diff
@@ -99,7 +99,44 @@ func CreateLockTable(tableName string, readCapacityUnits int, writeCapacityUnits
+
+func tagTableITagsGiven(tags map[string]string, createTableOutput *dynamodb.CreateTableOutput, client *dynamodb.DynamoDB, terragruntOptions *options.TerragruntOptions) error {
+
+	if tags == nil {
```
Check if `len(tags) == 0` too

BernhardBln replied:
Done

brikis98 commented on dynamodb/dynamo_lock_table.go:
```diff
@@ -99,7 +99,44 @@ func CreateLockTable(tableName string, readCapacityUnits int, writeCapacityUnits
+
+	var input = dynamodb.TagResourceInput{
+		ResourceArn: createTableOutput.TableDescription.TableArn,
+		Tags:        tagsConverted}
```
Make sure to run `make fmt`

BernhardBln replied:
Done

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -15,6 +15,13 @@ import (
 )
 
 // A representation of the configuration options available for S3 remote state
```
Comment needs to be moved to `RemoteStateConfigS3` and the comment here should explain that `ExtendedRemoteStateConfigS3` is used to store custom settings that are not used by Terraform, but by Terragrunt only.

BernhardBln replied:
Done

brikis98 commented on remote/remote_state.go:
```diff
@@ -122,6 +122,11 @@ func (remoteState *RemoteState) differsFrom(existingBackend *TerraformBackend, t
+
+		if key == "s3_bucket_tags" || key == "dynamotable_tags" {
+			continue
+		}
```
Instead of doing this, couldn't you use `remoteState.Config.remoteStateConfigS3`?

BernhardBln replied:
Nope, because we are in remote_state.go, not in remote_state_s3.go, where we have this object.

The cleanest would probably be to have the individual backends return the configuration.

BernhardBln replied:
And I just implemented that, because it does not make sense that the generic remote_state needs to know which keys to filter for remote_state_s3. I encapsulated this there now, and extended the `RemoteStateInitializer` interface by another method `GetTerraformInitArgs`.

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -117,8 +126,28 @@ func parseS3Config(config map[string]interface{}) (*RemoteStateConfigS3, error)
+
+	if err := mapstructure.Decode(config, &extendedConfig); err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
```
Isn't it enough to just do this? Doesn't it automatically fill in the `remoteStateConfigS3` portion?

BernhardBln replied:
No, that doesn't work. It does create the extendedConfig.remoteStateConfigS3 object, but all the fields are empty.

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -78,29 +85,31 @@ func (s3Initializer S3Initializer) NeedsInitialization(config map[string]interfa
 		return err
 	}
 
+	var s3Config = s3ConfigExtended.remoteStateConfigS3
```
Why the intermediate variable? Can't you use `s3ConfigExtended.Region`, `s3ConfigExtended.Endpoint`, etc below?

BernhardBln replied:
No, s3configExtended has these fields only:

* remoteStateConfigS3
* S3BucketTags
* DynamotableTags

If I didn't have the intermediate variable, I would have to use `s3ConfigExtended.remoteStateConfigS3.Region`, which I find less readable.

My goal was to separate the actual terraform backend config from the two extra keys that are just for terragrunt.

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -173,20 +212,71 @@ func checkIfVersioningEnabled(s3Client *s3.S3, config *RemoteStateConfigS3, terr
+
+func TagS3Bucket(s3Client *s3.S3, config *ExtendedRemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
+
+	if len(config.S3BucketTags) == 0 {
```
Check for `nil` too

BernhardBln replied:
done

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -173,20 +212,71 @@ func checkIfVersioningEnabled(s3Client *s3.S3, config *RemoteStateConfigS3, terr
+func TagS3Bucket(s3Client *s3.S3, config *ExtendedRemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
+
+	if len(config.S3BucketTags) == 0 {
+		terragruntOptions.Logger.Print("No tags for S3 bucket given.")
```
`Println`?

BernhardBln replied:
Ups, yeah.

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -173,20 +212,71 @@ func checkIfVersioningEnabled(s3Client *s3.S3, config *RemoteStateConfigS3, terr
+
+	var tagsConverted []*s3.Tag
+
+	var i = 0
```
What's `i` for?

BernhardBln replied:
I used that before I figured out there is an append function ;-) Removed

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -173,20 +212,71 @@ func checkIfVersioningEnabled(s3Client *s3.S3, config *RemoteStateConfigS3, terr
+	}
+
+	// There must be one entry in the list
+	var tags = config.S3BucketTags[0]
```
Why only `[0]`?

BernhardBln replied:
S3BucketTags can only be empty, or have one element.

If you add several config blocks like this,

```
    config {
      bucket =...

      s3_bucket_tags {
        owner = "team name"
      }

      s3_bucket_tags {
         service = "Terraform"
      }
    }
```

I'm returning this MultipleTagsDeclarations error (see at bottom of remote_state_s3.go file), because in that case, we would have to merge them together, and make sure tags are not set twice, which I found a bit too much and hence decided to allow only one tags declaration.

In case somebody specifies a tag twice in the same configuration, it seems the last value wins, hence no problem for us.

brikis98 replied:
If the syntax is:

```hcl
    config {
      bucket =...

      s3_bucket_tags = {
        owner = "team name"
      }

      s3_bucket_tags = {
         service = "Terraform"
      }
    }
```

Could you make `S3BucketTags` a map rather than a list of maps?

brikis98 commented on remote/remote_state_s3.go:
```diff
@@ -258,7 +349,12 @@ func createLockTableIfNecessary(s3Config *RemoteStateConfigS3, terragruntOptions
+	var tags map[string]string = nil
+	if len(tagsDeclarations) == 1 {
+		tags = tagsDeclarations[0]
+	}
```
Again, why only `[0]`?

BernhardBln replied:
See above

BernhardBln commented:
Done fixing remarks from review; will now start to write some tests.

BernhardBln commented:
Done with tests. Hope they are sufficient so far.

Please check again, @brikis98 - especially my solution with the `RemoteStateInitializer`, in order to move the filtering by s3-backend-specific keys from `remote_state` to `remote_state_s3`.

brikis98 commented on dynamodb/dynamo_lock_table.go:
```diff
@@ -99,7 +99,48 @@ func CreateLockTable(tableName string, readCapacityUnits int, writeCapacityUnits
+	}
+
+	if createTableOutput != nil && createTableOutput.TableDescription != nil && createTableOutput.TableDescription.TableArn != nil {
+		// Do not tag in case somebody else had created the table
```
Oh, that's a very interesting point. Checking `createTableOutput` makes sense; why check the description field?

BernhardBln replied:
That part crashed in a test, so I guessed it was necessary as well. Just to be on the safe side...

brikis98 replied:
Crashed in a test? Because you didn't check description? Do you have the stack trace?

BernhardBln replied:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa0 pc=0x132babc]

goroutine 190 [running]:
github.com/gruntwork-io/terragrunt/dynamodb.CreateLockTable(0xc420014cc0, 0x16, 0x0, 0x1, 0x1, 0xc42000e0c8, 0xc4201e0000, 0x0, 0x0)
        /Users/bernhardmercateo/go/src/github.com/gruntwork-io/terragrunt/dynamodb/dynamo_lock_table.go:108 +0x4dc
github.com/gruntwork-io/terragrunt/dynamodb.CreateLockTableIfNecessary(0xc420014cc0, 0x16, 0x0, 0xc42000e0c8, 0xc4201e0000, 0x0, 0x0)
        /Users/bernhardmercateo/go/src/github.com/gruntwork-io/terragrunt/dynamodb/dynamo_lock_table.go:48 +0x161
github.com/gruntwork-io/terragrunt/dynamodb.TestCreateLockTableConcurrency.func1(0xc420020f80, 0xc420014cc0, 0x16, 0xc42000e0c8, 0xc4201e0000, 0xc4201c43c0)
        /Users/bernhardmercateo/go/src/github.com/gruntwork-io/terragrunt/dynamodb/dynamo_lock_table_test.go:47 +0x89
created by github.com/gruntwork-io/terragrunt/dynamodb.TestCreateLockTableConcurrency
        /Users/bernhardmercateo/go/src/github.com/gruntwork-io/terragrunt/dynamodb/dynamo_lock_table_test.go:45 +0x1b9
exit status 2
```

Hm, you think if createTableOutput is not nil, I can rely on the other two fields to be filled and don't need to check?

brikis98 replied:
Oh, I misunderstood `TableDescription`; I thought it was an optional human-readable description, rather than a struct used to return data about the table such as its ARN. OK, this is probably fine then.

brikis98 commented on dynamodb/dynamo_lock_table_test.go:
```diff
@@ -81,7 +81,52 @@ func TestCreateLockTableIfNecessaryTableAlreadyExists(t *testing.T) {
+	}
+
+	var tags = make(map[string]string)
+	tags["team"] = "team A"
```
Can be simplified to:

```go
tags := map[string]string{"team": "team A"}
```

BernhardBln replied:
Done

brikis98 commented on dynamodb/dynamo_lock_table_test.go:
```diff
@@ -81,7 +81,52 @@ func TestCreateLockTableIfNecessaryTableAlreadyExists(t *testing.T) {
+}
+
+func assertTags(expectedTags map[string]string, tableName string, client *dynamodb.DynamoDB, t *testing.T) {
+	var description, err = client.DescribeTable(&dynamodb.DescribeTableInput{TableName: &tableName})
```
Idiomatic approach is `aws.String(tableName)` instead of `&tableName`.

BernhardBln replied:
Done

brikis98 commented on dynamodb/dynamo_lock_table_test.go:
```diff
@@ -81,7 +81,52 @@ func TestCreateLockTableIfNecessaryTableAlreadyExists(t *testing.T) {
+	withLockTableTagged(t, tags, func(tableName string, client *dynamodb.DynamoDB) {
+		assertCanWriteToTable(t, tableName, client)
+
+		assertTags(tags, tableName, client, t)
```
Nice tests 👍

brikis98 commented on remote/remote_state.go:
```diff
@@ -26,6 +26,8 @@ type RemoteStateInitializer interface {
 	// Initialize the remote state
 	Initialize(config map[string]interface{}, terragruntOptions *options.TerragruntOptions) error
+
+	GetTerraformInitArgs(config map[string]interface{}) map[string]interface{}
```
Please add a comment explaining what this method does and why it exists

BernhardBln replied:
Done

brikis98 commented on remote/remote_state.go:
```diff
@@ -121,7 +123,16 @@ func (remoteState *RemoteState) differsFrom(existingBackend *TerraformBackend, t
+	initializer, hasInitializer := remoteStateInitializers[remoteState.Backend]
+	if !hasInitializer {
+		// return empty array
+		return backendConfigArgs
```
Doesn't seem backwards compatible to return an empty array here. Instead, you should loop over `remoteState.Config` as before to fill `backendConfigArgs`. Or, more accurately, you should set `config = remoteState.Config` by default and then override it with `initializer.GetTerraformInitArgs` if `initializer` exists.

BernhardBln replied:
True, fixed and also wrote a test for that.

brikis98 commented on remote/remote_state_test.go:
```diff
@@ -18,10 +21,20 @@ func TestToTerraformInitArgs(t *testing.T) {
 	}
 	args := remoteState.ToTerraformInitArgs()
 
+	// must not contain s3_bucket_tags or dynamotable_tags
```
Does this test check the S3 tags were actually created?

BernhardBln replied:
No, it just calls ToTerraformInitArgs.

It seems there is no test at all that actually creates the bucket, if you guys add one later I'd be happy to add the check for the tags.

brikis98 replied:
A bunch of the tests in [integration_test.go](https://github.com/gruntwork-io/terragrunt/blob/master/test/integration_test.go) create an S3 bucket. See the fixtures folder.

BernhardBln replied:
Ok found and extended it to check for tags of both s3 and dynamo db.

The test case I changed runs on my machine, if I let the whole test run, I get this, but I assume it's not related to my changes:

```
[terragrunt] [/var/folders/3h/j7zf9_595vb97k60z03zt2nh0000gp/T/terragrunt-test258914403/fixture-output-all/env1/app1] 2018/06/15 11:42:21 Running command: terraform apply -input=false -auto-approve -var terraform_remote_state_s3_bucket="terragrunt-test-bucket-mgqpz9"
Copying configuration from "file:///Users/bernhard/go/src/github.com/gruntwork-io/terragrunt/test/fixture-download/hello-world"...
Initializing modules...
- module.hello
- module.remote
  Getting source "github.com/gruntwork-io/terragrunt.git//test/fixture-download/hello-world?ref=v0.9.9"
Error downloading modules: Error loading modules: subdir "test/fixture-download/hello-world" not found
--- FAIL: TestLocalDownloadWithHiddenFolder (0.14s)
	integration_test.go:817: Failed to run Terragrunt command 'terragrunt apply --terragrunt-non-interactive --terragrunt-working-dir fixture-download/local-with-hidden-folder' due to error: Hit multiple errors:
		exit status 1
```

BernhardBln commented:
@brikis98 - fixed everything and added a new test for the case that there is no backend available to return the config.

For some reason, I cannot reply to your comment about "Could you make S3BucketTags a map rather than a list of maps?", there is no "Reply" field. 

Is there a way to I can get this as a map, rather than slice of map?

No matter if I use

```
s3_bucket_tags = {
    service = "Terraform"
}
```

or

```
s3_bucket_tags {
    service = "Terraform"
}
```

I get them as slice of map: `'s3_bucket_tags' expected a map, got 'slice'`

BernhardBln commented:
I extended the integration test (see my other comment), renamed the config key `dynamotable_tags` to `dynamodb_table_tags` for the sake of consistency, and also added the two new config keys to the README.

Let me know if there is anything we can do about this map/slice of map problem, otherwise I think we should be pretty much done soon :)

brikis98 commented on README.md:
```diff
@@ -494,7 +504,9 @@ terragrunt = {
 as Terraform. The next time you run `terragrunt`, it will automatically configure all the settings in the
 `remote_state.config` block, if they aren't configured already, by calling [terraform
-init](https://www.terraform.io/docs/commands/init.html).
+init](https://www.terraform.io/docs/commands/init.html). The config options `s3_bucket_tags` and 
```
Nice! Thx for updating the docs :)

brikis98 commented on remote/remote_state_test.go:
```diff
@@ -18,10 +21,37 @@ func TestToTerraformInitArgs(t *testing.T) {
 	}
 	args := remoteState.ToTerraformInitArgs()
 
+	// no Backend initializer available, but command line args should still be passed on
```
👍

BernhardBln commented:
Sweet, thanks a lot, also for the patience :)

brikis98 commented:
Whoops, almost forgot to create the new release! Here it is (new binaries should show up in a few min): https://github.com/gruntwork-io/terragrunt/releases/tag/v0.14.11

BernhardBln commented:
Thank you :)
